File: java/src/com/pressel/toolkit/Wonder.java
/*
 * To change this license header, choose License Headers in Project Properties. To change this
 * template file, choose Tools | Templates and open the template in the editor.
 */

package com.pressel.toolkit;

/**
 *
 * @author Andrew Pressel
 */
public class Wonder {

  private String format;

  private long startTime;
  private int tick;
  private int total;
  private int frequency;
  private int threshold;
  private String header;

  public Wonder(final int total) {
    this(total, 20);
  }

  public Wonder(final int total, final int frequency) {
    this.total = total;
    this.frequency = frequency;
    reset();
  }

  public void reset() {
    startTime = System.currentTimeMillis();
    tick = 1;
    int wide = String.valueOf(total).length();
    format = "%" + wide + "d  %" + wide + "d  %5.01f  %12s  %8.01f  %12s\n";
    header = String.format("%" + wide + "s  %" + wide + "s  %5s  %12s  %8s  %12s\n", "Done",
        "Total", "%", "Elapsed", "Rate/s", "Estimate");

  }

  public void tick() {
    tick++;
    if ((tick % ((total / frequency) + 1)) == 0 || (tick == total)) {
      banner();
    }
  }

  public void banner() {
    System.out.print(header);
    header = "";
    long now = System.currentTimeMillis();
    long delta = (now - startTime);
    float rate = ((float) tick / delta);
    long estimate = (long) ((total - tick) / rate);
    // quick(rate, tick, rate / tick, total - tick, estimate);
    String row = String.format(format, tick, total, ((float) tick / total) * 100, timeString(delta),
        (rate * 1000), timeString(estimate));
    System.out.print(row);
  }

  private String timeString(long millis) {
    int hr = (int) (millis / 3600000);
    millis = millis % 3600000;
    int mn = (int) (millis / 60000);
    millis %= 60000;
    int sc = (int) (millis / 1000);
    millis %= 1000;

    return String.format("%02d:%02d:%02d.%03d", hr, mn, sc, millis);
  }

}

File: cpp/src/Wonder.cpp
#include "Wonder.h"

#include <iostream>
#include <iomanip>
#include <sstream>

Wonder::Wonder(int total, int frequency){
  _total = total;
  _frequency = frequency;
  reset();
}

Wonder::~Wonder(){}

void Wonder::reset(){
  _header = true;
  _start = duration_cast<milliseconds>(system_clock::now().time_since_epoch());
  _tick = 0;
  _wide = 0;
  int n = _total;
  while( n ){
    _wide++;
    n /= 10;
  }

}

void Wonder::tick(){
  _tick++;
  if( (_tick  % ((_total/_frequency)+1)) == 0 || (_tick == _total) ){
    banner();
  }
}

void Wonder::banner(){
  if( _header ){
    _header = false;
    std::cout
      << std::setw(_wide) << "Done" << "  "
      << std::setw(_wide) << "Total" << "  "
      << std::setw(5) << "%" << "  "
      << std::setw(12) << "Elapsed" << "  "
      << std::setw(8) << "Rate/s" << " "
      << std::setw(12) << "Estimate"
      << std::endl;
  }

  milliseconds now = duration_cast<milliseconds>(system_clock::now().time_since_epoch());
  long delta = (now.count() - _start.count());
  float rate = ((float)_tick / delta);
  long estimate = (long)((_total - _tick) / rate);

  std::cout
    << std::setw(_wide) << _tick << "  "
    << std::setw(_wide) << _total << "  "
    << std::setw(5) << std::fixed <<  std::setprecision(1) << ((float)_tick / _total) * 100 << "  "
    << std::setw(12) << timeString(delta) << "  "
    << std::setw(8) << (rate*1000) << " "
    << std::setw(12) << timeString(estimate)
    << std::endl;
}

std::string Wonder::timeString(long millis){
  int hr = (int)(millis / 3600000);
  millis = millis % 3600000;
  int mn = (int)(millis / 60000);
  millis %= 60000;
  int sc = (int)(millis / 1000);
  millis %= 1000;

  std::ostringstream buffer;
  buffer << std::setw(2) << std::setfill('0') << hr << ':'
    << std::setw(2) << std::setfill('0') << mn << ':'
    << std::setw(2) << std::setfill('0') << sc << '.'
    << std::setw(3) << std::setfill('0') << millis;
  return buffer.str();
}
